goodkey/good_key.go

```diff
@@ -351,3 +376,51 @@ func checkSmallPrimes(i *big.Int) bool {
 	result.GCD(nil, nil, i, smallPrimesProduct)
 	return result.Cmp(big.NewInt(1)) != 0
 }
+
+// Returns true if the modulus n is able to be factored into primes p and q via
+// Fermat's factorization method. This method relies on the two primes being
+// very close together, which means that they were almost certainly not picked
+// independently from a uniform random distribution. Basically, if we can factor
+// the key this easily, so can anyone else.
+func checkPrimeFactorsTooClose(n *big.Int, rounds int) bool {
+	// Pre-allocate some big numbers that we'll use a lot down below.
+	one := big.NewInt(1)
+	bb := big.NewInt(0)
+
+	// Any odd integer is equal to a difference of squares of integers:
+	//   n = a^2 - b^2 = (a + b)(a - b)
+	// Any RSA public key modulus is equal to a product of two primes:
+	//   n = pq
+	// Here we try to find values for a and b, since doing so also gives us the
+	// prime factors p = (a + b) and q = (a - b).
+
+	// We start with a close to the square root of the modulus n, to start with
+	// two candidate prime factors that are as close together as possible and
+	// work our way out from there. Specifically, we set a = ceil(sqrt(n)), the
+	// first integer greater than the square root of n. Unfortunately, big.Int's
+	// built-in square root function takes the floor, so we have to add one to get
+	// the ceil.
+	a := big.NewInt(0)
+	a.Sqrt(n).Add(a, one)
+
+	// We calculate b^2 to see if it is a perfect square, and therefore b is an
+	// integer. Specifically, b^2 = a^2 - n.
```

Comment:
```suggestion
	// We calculate b2 to see if it is a perfect square (b^2), and therefore b is an
	// integer. Specifically, b2 = a^2 - n.
```
"calculate b^2 to see if it is a perfect square" is trivially true. :-)

[The wikipedia article](https://en.wikipedia.org/wiki/Fermat%27s_factorization_method) seems to use "b2" (no caret) to indicate "a value that we hope is equal to b^2 for some b".